### PR TITLE
[DPE-3179] Use specified DB + Application roles from host charm

### PR DESCRIPTION
**Pull request by MiaAltieri**
# Problem
Right now the sharded cluster creates a DB for an outside application using a preset database name + user roles

# Solution
Retrieve the database name + user roles from the hosting application

### CODE REVIEW

**Review by dmitry-ratushnyy — DISMISSED** [review]

File: lib/charms/mongos/v0/mongos_client_interface.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        database_name: str,
+        extra_user_roles: str,
+        relation_name: str = MONGOS_RELATION_NAME,
+    ) -> None:
+        """Constructor for MongosRequirer object."""
+        self.relation_name = relation_name
+        self.charm = charm
+
+        if not database_name:
+            database_name = f"{self.charm.app}-mongos"
```
Comment (Mehdi-Bendriss) [inline lib/charms/mongos/v0/mongos_client_interface.py:123]: What is the use of having a `-mongos` suffix? to indicate that this database was created in a sharded deployment? if yes, do you think it's really needed?

File: src/charm.py [hunk 2]
```diff
@@ -237,9 +239,9 @@ def share_uri(self) -> None:
         # TODO future PR - generate different URI for data-integrator as that charm will not
         # communicate to mongos via the Unix Domain Socket.
 
-    def set_user_role(self, roles: List[str]):
+    def set_user_role(self, roles: List[str]) -> None:
```
Comment (Mehdi-Bendriss) [inline src/charm.py:242]:
```suggestion
    def set_user_roles(self, roles: List[str]) -> None:
```

File: tests/integration/helpers.py [hunk 3]
```diff
@@ -10,20 +10,49 @@
...
 
+async def run_mongos_command(ops_test: OpsTest, unit: ops.model.Unit, mongos_cmd: str):
+    """Runs the provided mongos command.
+
+    The mongos charm uses the unix domain socket to communicate, and therefore we cannot run
+    MongoDB commands from outside the unit and we must use `juju exec` instead.
+    """
+    mongodb_uri = await generate_mongos_uri(ops_test, auth=True)
+
+    check_cmd = [
```
Comment (Mehdi-Bendriss) [inline tests/integration/helpers.py:42]: nit: an alternative approach may be to write the command as a string with an f-string placeholder, and then split it in the `.juju(..)` call

**Review by Mehdi-Bendriss — DISMISSED** [review]
Great work! I only have a question and a couple of nits

File: lib/charms/mongos/v0/mongos_client_interface.py (hunk 1 above)
Comment (MiaAltieri) [reply lib/charms/mongos/v0/mongos_client_interface.py:123]: Yes exactly! But after further reflecition I guess you're right, not technically needed. I'll remove it!

File: tests/integration/helpers.py (hunk 3 above)
Comment (MiaAltieri) [reply tests/integration/helpers.py:42]: Would be nice in theory, but the `.juju` call doesn't accept the input as a list, instead commands, which is why it is needed to create a list, and then reference it in the command as `.juju(*check_cmd)`

**Review by Mehdi-Bendriss — DISMISSED** [review]

**Review by dmitry-ratushnyy — DISMISSED** [review]

**Review by dmitry-ratushnyy — APPROVED** [review]

**Review by Mehdi-Bendriss — APPROVED** [review]

**Pull request merged**